### DISCUSSION TITLE
WIP - StatusNotifier widget

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#2580] [Show and tell]
Thought it might be good to show the progress that I'm making on the StatusNotifier widget. Unlike the SysTray widget that only works on X11 backends, this should also work on Wayland (but there's a small bug at the moment...).

Currently, the widget looks like this:

https://user-images.githubusercontent.com/946265/123508309-7bfb3100-d666-11eb-81cb-d20a4f5eefda.mp4

The context menus are built using the popup toolkit I built in #2251.

There's still a bit to do:
- [x] Fix popup creation error on Wayland (this could be a bigger issue than I hoped)
- [x] Better handling of blank icons
- [x] Handling of icons with no context menu
- [x] Use installed icons in preference to pixmaps (may want to use pyxdg to simplify loading icons)
- [ ] Checking which icon to use (normal vs "attention")
- [x] Listen for updating properties on DBus (shouldn't be needed as we can listen for specific signals)
- [x] Better positioning of context menu
- [ ] When to close menu if no item selected (e.g. clicking elsewhere)

But we're close

**Comment by elParaguayo** [comment]
@m-col This is the issue I'm having with Wayland at the moment: clicking on the popup menu seems to send the button clicks to an entirely different widget (text box showing the coordinated and button number)


https://user-images.githubusercontent.com/946265/124102393-3d93b680-da58-11eb-9112-ba2ef7eecb67.mp4

**Reply by m-col** [reply]
Even with https://github.com/qtile/qtile/commit/a41734d0d9b207af93646a75dc92966ee7cc7b69?

Though I see not only is this a different widget, it's a whole different window! Could you try to see if you see the same thing if the windows do not overlap, so that the cursor moves off one `Internal`, to the desktop, and then onto the other `Internal`? The wayland `Core` tracks which internal window has pointer focus in `self._hovered_internal` and I wonder if it is not getting updated correctly if the cursor moves from one to another directly without a gap.

**Reply by elParaguayo** [reply]
Yes, it includes a41734d. Before that, the coordinates it gave me were even more off.

You've nailed it with the overlapping windows. If I have a space between the bar and the popup then it seems to work. Clickling "Quit" on the menu closed vlc. However, I'm not getting any of my mouse over highlighting but I can almost certainly chase that one down.

**Reply by m-col** [reply]
Does this fix it?
```diff
diff --git a/libqtile/backend/wayland/core.py b/libqtile/backend/wayland/core.py
index 4bca8b3a..7a5a326a 100644
--- a/libqtile/backend/wayland/core.py
+++ b/libqtile/backend/wayland/core.py
@@ -380,7 +380,7 @@ class Core(base.Core, wlrq.HasListeners):
         if found:
             win, surface, sx, sy = found
             if isinstance(win, window.Internal):
-                if self._hovered_internal:
+                if self._hovered_internal is win:
                     win.process_pointer_motion(self.cursor.x, self.cursor.y)
                 else:
                     self.seat.pointer_clear_focus()

```

**Reply by elParaguayo** [reply]
Yes. That puts the mouse clicks to the right window, even if they're overlapping.

**Reply by m-col** [reply]
Does it also fix the mouse-over highlighting? This is just a pointer motion response, right?

**Reply by m-col** [reply]
Cool wallpaper btw - is that from KDE?

**Reply by elParaguayo** [reply]
No, doesn't fix the mouse-over - I'll need to do some debugging on that. Probably something to do with how it's being drawn.

Wallpaper is my own. Made it in Inkscape.

**Reply by m-col** [reply]
RIght, gotcha.

> Wallpaper is my own. Made it in Inkscape.

Nice!

**Reply by elParaguayo** [reply]
The mouse motion response isn't working because it's still getting absolute x and y coordinates, rather than being relative to the window.

**Reply by elParaguayo** [reply]
Think it just needs this:
``` diff
diff --git a/libqtile/backend/wayland/core.py b/libqtile/backend/wayland/core.py
index 4bca8b3a..ce248fd6 100644
--- a/libqtile/backend/wayland/core.py
+++ b/libqtile/backend/wayland/core.py
@@ -380,8 +380,11 @@ class Core(base.Core, wlrq.HasListeners):
         if found:
             win, surface, sx, sy = found
             if isinstance(win, window.Internal):
-                if self._hovered_internal:
-                    win.process_pointer_motion(self.cursor.x, self.cursor.y)
+                if self._hovered_internal is win:
+                    win.process_pointer_motion(
+                        self.cursor.x - self._hovered_internal.x,
+                        self.cursor.y - self._hovered_internal.y
+                    )
                 else:
                     self.seat.pointer_clear_focus()
                     win.process_pointer_enter(self.cursor.x, self.cursor.y)
```

I have highlighting working after that.

Do you want to put your fixes in a separate commit?

**Reply by m-col** [reply]
Nah, just stick em together
